modules/diskq/dqtool.c

```diff
@@ -177,6 +203,343 @@ dqtool_info(int argc, char *argv[])
   return 0;
 }
 
+static gboolean
+_is_read_writable(const gchar *path)
+{
+  return access(path, R_OK|W_OK) == F_OK;
+}
+
+static gboolean
+_validate_diskq_path(const gchar *path)
+{
+  if (!path)
+    {
+      fprintf(stderr, "relocate: missing mandatory option: new_path\n");
+      return FALSE;
+    }
+
+  if (!(is_file_directory(path) && _is_read_writable(path)))
+    {
+      fprintf(stderr, "relocate: new_path should be point to a readable/writable directory\n");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_validate_persist_file_path(const gchar *path)
+{
+  if (!path)
+    {
+      fprintf(stderr, "relocate: missing mandatory option: persist\n");
+      return FALSE;
+    }
+
+  if (!(is_file_regular(path) && _is_read_writable(path)))
+    {
+      fprintf(stderr, "relocate: persist should point to a readable/writable file\n");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_relocate_validate_options(void)
+{
+  return _validate_diskq_path(new_diskq_path) && _validate_persist_file_path(persist_file_path);
+}
+
+static gint
+_open_fd_for_reading(const gchar *fname)
+{
+  gint fd = open(fname, O_RDONLY);
+
+  if (fd == -1)
+    fprintf(stderr, "Failed to open %s, reason: %s\n", fname, strerror(errno));
+
+  return fd;
+}
+
+static gboolean
+_get_content_length_by_fd(gint fd, gsize *size)
+{
+  struct stat st;
+
+  if (fstat(fd, &st) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      *size = 0;
+      return FALSE;
+    }
+
+  *size = st.st_size;
+  return TRUE;
+}
+
+static gboolean
+_mmap_content_from_fd(gint fd, gchar **content, gsize *content_length)
+{
+  if (!_get_content_length_by_fd(fd, content_length))
+    return FALSE;
+
+  *content = mmap(NULL, *content_length, PROT_READ, MAP_PRIVATE, fd, 0);
+  if (*content == MAP_FAILED)
+    {
+      fprintf(stderr, "mmap failed, reason: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_mmap_copy_to_fd(gint fd, gchar *content, gsize content_length)
+{
+  if (ftruncate(fd, content_length) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  gchar *dst_content = mmap(NULL, content_length, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+  if (dst_content == MAP_FAILED)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  memcpy(dst_content, content, content_length);
+
+  if (msync(dst_content, content_length, MS_SYNC) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gint
+_open_fd_for_writing(const gchar *fname)
```

Comment:
[Optional]: The copy related functions clutter this file a little bit. I suggest adding them to a separate c file.